File: Kotlin/app/src/main/java/com/example/tictactoe/GameScreen.kt
package com.example.tictactoe

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import kotlin.random.Random

@Composable
fun GameScreen(onBack: () -> Unit, onGameEnd: (String) -> Unit, gameSettings: GameSettings) {
    var currentPlayer by remember { mutableStateOf(Player.X) }
    var board by remember {
        mutableStateOf(
            List(gameSettings.boardSize) { MutableList(gameSettings.boardSize) { Player.NONE } }
        )
    }
    var winner by remember { mutableStateOf<Player?>(null) }

    val player1Color = gameSettings.player1Color
    val player2Color = gameSettings.player2Color
    val player1Shape = gameSettings.player1Shape.symbol
    val player2Shape = gameSettings.player2Shape.symbol

    fun getPlayerSymbolAndColor(player: Player): Pair<String, Color> {
        return when (player) {
            Player.X -> player1Shape to player1Color
            Player.O -> player2Shape to player2Color
            Player.NONE -> "" to Color.Transparent
        }
    }

    fun checkWinner(): Player? {
        for (i in 0 until gameSettings.boardSize) {
            if ((board[i].all { it == Player.X }) || (board.map { it[i] }.all { it == Player.X })) return Player.X
            if ((board[i].all { it == Player.O }) || (board.map { it[i] }.all { it == Player.O })) return Player.O
        }
        if ((0 until gameSettings.boardSize).all { board[it][it] == Player.X }) return Player.X
        if ((0 until gameSettings.boardSize).all { board[it][it] == Player.O }) return Player.O
        if ((0 until gameSettings.boardSize).all { board[it][gameSettings.boardSize - it - 1] == Player.X }) return Player.X
        if ((0 until gameSettings.boardSize).all { board[it][gameSettings.boardSize - it - 1] == Player.O }) return Player.O
        return if (board.flatten().all { it != Player.NONE }) Player.NONE else null
    }

    fun makeComputerMove() {
        val emptyCells = board.flatMapIndexed { i, row -> row.mapIndexed { j, cell -> i to j }.filter { board[it.first][it.second] == Player.NONE } }
        if (emptyCells.isNotEmpty()) {
            val (i, j) = emptyCells.random()
            board = board.toMutableList().apply {
                this[i] = this[i].toMutableList().apply { this[j] = Player.O }
            }
            currentPlayer = Player.X
        }
    }

    fun onCellClick(i: Int, j: Int) {
        if (board[i][j] == Player.NONE && winner == null) {
            board = board.toMutableList().apply {
                this[i] = this[i].toMutableList().apply { this[j] = currentPlayer }
            }
            winner = checkWinner()
            if (winner == null) {
                currentPlayer = if (currentPlayer == Player.X) Player.O else Player.X
                if (gameSettings.mode == GameMode.SINGLE_MODE && currentPlayer == Player.O) {
                    makeComputerMove()
                    winner = checkWinner()
                }
            }
        }
    }

    // Wywołanie efektu po zmianie zwycięzcy
    LaunchedEffect(winner) {
        if (winner != null) {
            kotlinx.coroutines.delay(100)
            onGameEnd(
                if (winner == Player.NONE) {
                    "It's a tie!"
                } else {
                    val winnerShape = if (winner == Player.X) player1Shape else player2Shape
                    "The winner is: $winnerShape"
                }
            )
        }
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Color.Black),
        verticalArrangement = Arrangement.SpaceBetween,
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Box(
            modifier = Modifier.weight(1f),
            contentAlignment = Alignment.Center
        ) {
            Column {
                for (i in 0 until gameSettings.boardSize) {
                    Row {
                        for (j in 0 until gameSettings.boardSize) {
                            val cellState = board[i][j]
                            val (symbol, color) = getPlayerSymbolAndColor(cellState)
                            Box(
                                modifier = Modifier
                                    .size(80.dp)
                                    .border(2.dp, Color.Yellow)
                                    .clickable { onCellClick(i, j) },
                                contentAlignment = Alignment.Center
                            ) {
                                Text(
                                    text = symbol,
                                    style = MaterialTheme.typography.displayMedium,
                                    color = color
                                )
                            }
                        }
                    }
                }
            }
        }

        Button(
            onClick = onBack,
            modifier = Modifier
                .padding(32.dp)
        ) {
            Text("Back to settings")
        }
    }
}



enum class Player(val symbol: String) {
    X("X"), O("O"), NONE("")
}

File: react_native/tictactoe/screens/GameScreen.js
import React, { useState, useEffect } from 'react';
import { View, Text, TouchableOpacity } from 'react-native';
import Styles from "../ui.theme/Styles";

const GameScreen = ({ navigation, route }) => {
    const { gameSettings } = route.params;
    const styles = Styles();

    const [currentPlayer, setCurrentPlayer] = useState('X');
    const [board, setBoard] = useState(
        Array(gameSettings.boardSize).fill().map(() => Array(gameSettings.boardSize).fill('NONE'))
    );
    const [winner, setWinner] = useState(null);

    const player1Color = gameSettings.player1Color;
    const player2Color = gameSettings.player2Color;
    const player1Shape = gameSettings.player1Shape?.symbol || 'X';
    const player2Shape = gameSettings.player2Shape?.symbol || 'O';

    const getPlayerSymbolAndColor = (player) => {
        switch (player) {
            case 'X':
                return { symbol: player1Shape, color: player1Color };
            case 'O':
                return { symbol: player2Shape, color: player2Color };
            default:
                return { symbol: '', color: 'transparent' };
        }
    };

    useEffect(() => {
        if (!board.length) {
            setBoard(
                Array(gameSettings.boardSize).fill().map(() => Array(gameSettings.boardSize).fill('NONE'))
            );
        }
    }, []);

    const checkWinner = () => {
        for (let i = 0; i < gameSettings.boardSize; i++) {
            if (
                board[i].every(cell => cell === 'X') ||
                board.map(row => row[i]).every(cell => cell === 'X')
            ) return 'X';
            if (
                board[i].every(cell => cell === 'O') ||
                board.map(row => row[i]).every(cell => cell === 'O')
            ) return 'O';
        }
        if (board.every((_, i) => board[i][i] === 'X')) return 'X';
        if (board.every((_, i) => board[i][i] === 'O')) return 'O';
        if (board.every((_, i) => board[i][gameSettings.boardSize - i - 1] === 'X')) return 'X';
        if (board.every((_, i) => board[i][gameSettings.boardSize - i - 1] === 'O')) return 'O';
        if (board.flat().every(cell => cell !== 'NONE')) return 'NONE';
        return null;
    };

    const makeComputerMove = () => {
        const emptyCells = [];
        for (let i = 0; i < gameSettings.boardSize; i++) {
            for (let j = 0; j < gameSettings.boardSize; j++) {
                if (board[i][j] === 'NONE') emptyCells.push([i, j]);
            }
        }
        if (emptyCells.length) {
            const [i, j] = emptyCells[Math.floor(Math.random() * emptyCells.length)];
            board[i][j] = 'O';
            //setBoard([...board]);
            setCurrentPlayer('X');
        }
    };

    const onCellClick = (i, j) => {
        if (board[i][j] === 'NONE' && winner === null) {
            board[i][j] = currentPlayer;
            //setBoard([...board]);
            const currentWinner = checkWinner();
            setWinner(currentWinner);
            if (currentWinner === null) {
                setCurrentPlayer(currentPlayer === 'X' ? 'O' : 'X');
                if (gameSettings.mode === 'SINGLE_MODE' && currentPlayer === 'O') {
                    makeComputerMove();
                    const newWinner = checkWinner();
                    setWinner(newWinner);
                }
            }
        }
    };

    useEffect(() => {
        if (winner !== null) {
            const resultMessage = winner === 'NONE' ? "It's a tie!" : `The winner is: ${winner === 'X' ? player1Shape : player2Shape}`;
            setTimeout(() => {
                navigation.navigate('Result', { resultMessage, gameSettings: gameSettings });
            }, 100);
        }
    }, [winner]);

    return (
        <View style={styles.gameContainer}>
            <View style={styles.board}>
                {Array.from({ length: gameSettings.boardSize }).map((_, i) => (
                    <View key={i} style={styles.boardRow}>
                        {Array.from({ length: gameSettings.boardSize }).map((_, j) => {
                            const { symbol, color } = getPlayerSymbolAndColor(board[i][j]);
                            return (
                                <TouchableOpacity
                                    key={j}
                                    style={[styles.cell]}
                                    onPress={() => onCellClick(i, j)}
                                >
                                    <Text style={[styles.cellText, { color: color }]}>
                                        {symbol}
                                    </Text>
                                </TouchableOpacity>
                            );
                        })}
                    </View>
                ))}
            </View>

            <TouchableOpacity
                style={[styles.backButton]}
                onPress={() => navigation.goBack()}
            >
                <Text style={styles.playButtonText}>Back to settings</Text>
            </TouchableOpacity>

        </View>
    );
};

export default GameScreen;
